server/bin/pbench-backup-tarballs.py

```diff
@@ -0,0 +1,448 @@
+                "shutil.copy: Unable to copy {} from archive to backup: {}\n".format(archive_md5, backup_controller_path))
+        else:
+            md5_done = True
+
+        # copy the tarball from archive to backup
+        if md5_done:
+            try:
+                shutil.copy(tar, backup_controller_path)
+            except Exception:
+                # couldn't copy tarball
+                tar_done = False
+                logger.exception(
+                    "shutil.copy: Unable to copy {} from archive to backup: {}\n".format(tar, backup_controller_path))
+
+                # remove the copied md5 file from backup
+                bmd5_file = os.path.join(
+                    backup_controller_path, "{}.md5".format(resultname))
+                if os.path.exists(bmd5_file):
+                    try:
+                        os.remove(bmd5_file)
+                    except Exception:
+                        logger.exception("Unable to remove: {}".format(bmd5_file))
+            else:
+                tar_done = True
+
+        if md5_done and tar_done:
+            logger.info(
+                "Locally Backed-up Sucessfully: {}/{}\n".format(controller, resultname))
+            return Status.SUCCESS
+        else:
+            return Status.FAIL
+
+
+def backup_to_s3(s3_obj, logger, controller_path, controller, tb, tar, resultname, archive_md5_hex_value):
+    if s3_obj is None:
+        # Short-circuit operation when we don't have an S3 object to work with
+        # when executing.  This can happen when the expected bucket does not
+        # exist, or for other errors where we still want to backup locally.
+        return Status.FAIL
+
+    s3_resultname = os.path.join(controller, resultname)
+
+    # Check if the result already present in s3 or not
+    try:
+        obj = s3_obj.connector.get_object(Bucket='{}'.format(
+            s3_obj.bucket_name), Key='{}'.format(s3_resultname))
+        in_s3 = True
+    except Exception:
+        in_s3 = False
+
+    if in_s3:
+        # compare md5 which we already have so no need to recalculate
+        s3_md5 = obj['ETag'].strip("\"")
+
+        if archive_md5_hex_value == s3_md5:
+            # declare success
+            logger.info(
+                "The tarball {} is already present in S3 bucket, with same md5\n".format(s3_resultname))
+            return Status.SUCCESS
+        else:
+            logger.error(
+                "The tarball {} is already present in S3 bucket, but with different MD5\n".format(s3_resultname))
+            return Status.FAIL
+    else:
+        md5_base64_value = (base64.b64encode(
+            bytes.fromhex(archive_md5_hex_value))).decode()
+        try:
+            with open(tar, 'rb') as data:
+                try:
+                    s3_obj.connector.put_object(
+                        Bucket=s3_obj.bucket_name, Key=s3_resultname, Body=data, ContentMD5=md5_base64_value)
+                except Exception as e:
```

Comment:
In light of the errors we got in testing, this is way too general: we need to distinguish bad md5 exceptions from EntityTooLarge and connection closed problems. We should retry the latter I think, but for EntityTooLarge, we'll have to do a multipart upload, per[ the doc that you found](https://docs.aws.amazon.com/AmazonS3/latest/dev/UploadingObjects.html).